python/res/job_queue/job_queue_node.py

```diff
@@ -90,21 +93,32 @@ def _job_monitor(self, driver, max_submit):
             self.run_done_callback()
         elif self.status == JobStatusType.JOB_QUEUE_EXIT:
             if self.submit_attempt < max_submit:
-                self._set_status(JobStatusType.JOB_QUEUE_WAITING)
-                self.started = False
+                self.thread_status = ThreadStatusType.THREAD_READY
+                return
             else:
                 self._set_status(JobStatusType.JOB_QUEUE_FAILED)
                 self.run_exit_callback()
         elif self.status == JobStatusType.JOB_QUEUE_IS_KILLED:
             pass
         else:
+            self.thread_status = ThreadStatusType.THREAD_FAILED
             raise AssertionError("Unexpected job status type after running job: {}".format(self.status))
 
+        self.thread_status = ThreadStatusType.THREAD_DONE
+
     def run(self, driver, max_submit=2):
-        self.started = True
-        x = Thread(target=self._job_monitor, args=(driver, max_submit))
-        x.start()
-        return x
+        # Prevent multiple threads working on the same object
+        if self.thread is not None and self.thread.is_alive():
```

Comment:
Is the later condition necessary?

Reply:
unless the thread has been started, you should not wait for it... could, in theory, lock the whole thing.

Reply:
I agree :+1: